### DISCUSSION TITLE
Implement PRESTO_EXPAND_DATA for trino

### DISCUSSION THREAD

**Discussion by giftig** [apache/superset#25713] [Ideas]
Following on from a discussion I had with @villebro a couple of months ago, the `PRESTO_EXPAND_DATA` logic was dropped from the trino engine spec in 2.1 as part of trying to separate trino and presto implementations and separate the trino engine spec from some hard to maintain features, with this one being high on the list due to lack of test coverage and the obscurity of the low-level logic it's doing.

My team relies on this feature as it's nigh impossible for users to build visualisations from nested data in trino without it. I'm raising this as a discussion point now as it's become a priority for us to bring this feature up to scratch and contribute a version of it built for trino, complete with better test coverage and documentation to make sure this feature can be well understood and supported going forwards this time. I'm starting to analyse the feature now to try to do that.

From a brief preliminary investigation, my current understanding is that the complex low-level logic was needed because presto/trino lets you query nested `ROW` fields by doing `row.level_1.level_2.level_3` for example, but doesn't provide a convenient way to discover those nested sub-fields in order to define the expanded columns on datasets; that prevents you building charts off of them, or seeing them expanded into individual columns / autocompleting them when in SQL Lab. The current logic in the presto spec is doing some fairly complex parsing to figure out the column definitions by analysing the ROW fields.

So firstly I'm going to see if anything's changed and if trino can help us expand out these nested column definitions without having to write something very complex to do it ourselves. And secondly, however simple or complex the current solution against trino ends up being, I'll write something a bit more encapsulated, unit tested and documented to achieve this for trino.

If any trino users can offer any thoughts on this, whether this is a feature others need, suggestions on how it might be possible, or insights into how the original presto version works, I'd be happy to hear them.

CC from the [rolodex](https://docs.google.com/spreadsheets/d/1U1qxiLvOX0kBTUGME1AHHi6Ywel6ECF8xk_Qy-V9R8c/edit#gid=0) in case this is of interest: @john-bodley @sujiplr @nytai @cccs-tom @villebro (and @rusackas)

**Comment by cccs-tom** [comment]
Fwiw, we get around this by automatically generating our virtual datasets using an exhaustive `SELECT` clause instead of a `SELECT *`. So we end up with a really long definition, but it works! It's a pain to maintain manually, but that's why we automate it and run it on a schedule.
```sql
SELECT
    root.branch.leaf1 AS alias1,
    root.branch.leaf2 AS alias2,
[...]
```

**Reply by giftig** [reply]
Ah, that's interesting. I think essentially you're doing the same thing this feature does, so if you're happy to share how you're generating those queries I think it's basically what we need to do in `expand_data`. I'm hoping I can get this information directly from trino somehow, but the fact that the logic for this expand data feature in presto is [a couple of hundred lines of fairly complex parsing](https://github.com/apache/superset/blob/master/superset/db_engine_specs/presto.py#L881) doesn't give me a lot of hope. Curious to see if yours is similar.

**Reply by cccs-tom** [reply]
Unfortunately, our setup is pretty specific to our stack. Our data producers are other teams within the same org, so we have a "contract" for how they write tables - they add table properties that give us hints for how to generate the datasets. Then our Iceberg catalogs are ingested into DataHub (some of our data producers also use dbt, which also have a nice DataHub integration). We treat DataHub as our "source of truth", so our script queries it to find out the schema of our datasets. This is all still a little new and not-quite-prod-yet, but we're getting there! :)

**Reply by giftig** [reply]
Ah, interesting. Thanks for sharing, always useful to hear how others are using these tools.

**Comment by giftig** [comment]
So it looks like the first problem is a structural type we get back from trino looks a bit like this, and we need to parse it ourselves:

```
row(foo varchar, bar varchar, deeper row(deeperfoo varchar, deeperbar varchar))
```

That's slightly annoying, but not an extremely difficult problem. I can see this is partly what `_parse_structural_column` is trying to do on the presto spec, but it's mixed in with some other logic and does it in a slightly weird way, e.g. several nested for loops and a lot of unusual splits instead of trying to tokenise or some other approach. It's also taking a `result` and mutating it instead of returning something.

The biggest problems with this logic are lack of testing, lack of separation of high- and low-level logic, and lack of separation of the problem in general. I've started drafting a replacement which just solves the first problem: parsing this into structural types so that we can read them more easily. I've made a new module `superset.db_engine_specs.type_parsing` for this and introduced some classes to give it some structure:

```python
ARRAY = "ARRAY"
ROW = "ROW"


class Type:
    """Wraps the concept of a type, allowing describing structural types"""
    def __init__(self, type: str):
        self.type = type


class Row(Type):
    """Represents a row type, containing named, typed children"""
    def __init__(self, children: list[Type]):
        super().__init__(ROW)
        self.children = children


class Array(Type):
    """
    Represents an array type, i.e. containing an arbitrary number of unnamed
    children of a single type
    """
    def __init__(self, child_type: Type):
        super().__init__(ARRAY)
        self.child_type = child_type


class TypeParser:
    """Superclass for type parsers, containing database-specific logic"""
    def parse_type(self, typedef: str) -> Type:
        """
        Parse a potentially complex type and return a structured type
        representing a full breakdown of the type. For a simple type like
        varchar this will just return a simple wrapper around the original
        string, whereas for a complex type like row(foo varchar, bar varchar)
        we'll produce a Row type which models the child types nested beneath.
        """
        return Type(typedef)
```

and a quick first attempt at making this parser work by recursing into the structure:

```python
from superset.db_engine_specs.type_parsing.base import Array, Row, Type, TypeParser

class TrinoTypeParser(TypeParser):
    """Parse ROW and ARRAY types into a structured type definition"""

    def _parse_row(self, typedef: str) -> Row:
        columns = [
            self.parse_type(c.strip()) for c in typedef[len("row("):-1].split(",")
        ]
        return Row(columns)

    def _parse_array(self, typedef: str) -> Array:
        inner_type = self.parse_type(typedef[len("array("):-1])
        return Array(inner_type)

    def parse_type(self, typedef: str) -> Type:
        if typedef.lower().startswith("row("):
            return self._parse_row(typedef)

        if typedef.lower().startswith("array("):
            return self._parse_array(typedef)

        return Type(typedef)
```

Once we have that structured information it's easier to do the next steps.

**Comment by giftig** [comment]
I guess my first question is: do I open a PR as I build each component we'll need, or does the PR need to be a fully-formed feature? I think it'll be much easier to review the building blocks if I start with this parsing + a full suite of unit tests for it and then build the next parts and raise additional PRs, but it won't be a working feature until all parts are in place.

I'll operate on the assumption for now that multiple small PRs are ok and raise one when I'm happy with this piece; it can always be discussed further on the PR.

**Comment by giftig** [comment]
I've raised a PR for part one: https://github.com/apache/superset/pull/25746 which is the most complicated part imo, parsing the type info trino provides into a structure we can then navigate. There were some challenges and limitations which I've tried to thoroughly document.

tl;dr = trino is sometimes ambiguous when spaces or brackets appear in ROW field names, and we can't do anything about it.

I'll start trying to piece this into the actual engine spec and create an `expand_data` implementation next.

**Comment by giftig** [comment]
One thing I've noticed while analysing the `PRESTO_EXPAND_DATA` feature is that as well as expanding `ROW` fields out into dotted path strings, it also has an additional visual effect in the results table, in SQL Lab only, which expands `ARRAY` fields out onto additional rows:

![screenshot_2023-10-20_16-20-26_0d717987](https://github.com/apache/superset/assets/2862039/802b69d6-2708-4fd5-91f6-56990a32ca73)

I think this is misleading for a few reasons:
  - it implies association between values 1 and 4, 2 and 5, and 3 and 6 in my example above, but they're not associated at all, they're elements with the same index in completely distinct arrays.
  - it becomes very unclear which rows are associated with which parent, especially with many heterogenous lists and many rows
  - it's not accurately representing the data -- arrays within a row are not additional rows, and can not be queried as if they are
  - this feature is a visual quirk only, and only appears in SQL Lab -- it has no logical equivalent when building charts and doesn't appear in table charts for example.

So I'm intending to build the trino version to expand out the columns into dotted paths as before, since these represent valid syntax in querying those fields in trino, but drop the expansion of arrays across additional rows.

**Comment by giftig** [comment]
Unfortunately I've looked closer at what `expand_data` takes in and outputs, and despite the docstring it's not getting structured data in, it's just getting raw data in like this:

```python
[{'field1': '[[1, 2, 3], [4, 5, 6], [["foo", "bar"], ["hodor", "HODOR HODOR HODOR"]], [["baz", "buzz"], ["bzzzzzzt", "BZZZZZT!"]]]', 'field2': 'toplevel'}]
```
and parsing that string data into structured data like this, prior to actioning the expansion into extra fields:
```python
{'field1': [[1, 2, 3],
             [4, 5, 6],
             [['foo', 'bar'], ['hodor', 'HODOR HODOR HODOR']],
             [['baz', 'buzz'], ['bzzzzzzt', 'BZZZZZT!']]]}
```
the parsing logic here is simply too fragile; I don't want to try to reproduce it. It's not aware of trino types, for one thing, which explains why previously I've had to fudge types in other parts of the trino db spec to fix unexpected python types rather than sqla types being provided, with this feature enabled.

This stops being a problem in charts, because we produce an exhaustive SELECT statement to power the charts, like the way @cccs-tom described.

I think the only sensible way we can present the table in the above format in SQL Lab is to do the same thing in SQL Lab; use the type information we've parsed to wrap the user's query in an exhaustive select for all inner fields, and that would get the structured data from trino rather than a blob we have to try to manually parse. I'll take a look into how we can do that in a reasonable way. It'll probably mean a change to the API on db engine spec since `expand_data` isn't going to the correct context; it's invoked after the query has already run.

If that proves excessively difficult, as a compromise it may make sense to do what we can with the type information and abandon the attempt to expand out a `SELECT *` into separate columns in the SQL Lab results table. With just the type information we can at least get a more comprehensive table schema summary, autocomplete on dotted path fields in SQL Lab, and expand out the fields in virtual datasets created from `SELECT *` queries, which is a big improvement on the status quo. I could also polish off that part at this point and then put in additional effort to get the data working as a further improvement, since they're relatively distinct features.

**Comment by giftig** [comment]
Another thing I've noticed, and I suppose this explains why we seem to be the only one who noticed `PRESTO_EXPAND_DATA` got dropped from trino: it doesn't actually work for previews at all, it just produces an error. We've had a query mutator in our config which fixes it for quite a while. `PRESTO_EXPAND_DATA` causes `get_columns` to add columns with names like `foo.bar.baz`. The base spec logic then converts columns into sqla query select `Column`s by simply wrapping them in `column(col["name"])`.

This will produce something like this:

```sql
SELECT "foo.bar" FROM table
```

which doesn't work. Our mutator has long since done some fiddly string parsing to turn that into this instead;

```sql
SELECT "foo"."bar" AS "foo.bar" FROM table
```
which correctly queries and labels it. I'm now experimenting with adding (yet another) optional  field to `ResultSetColumnType` to allow this information to be conveyed to the base spec, and altering the base spec's `_get_fields` logic to respect it where provided; that should fix that issue.

**Reply by giftig** [reply]
I wasn't optimistic it'd be that easy but it is; that solves the problem nicely. I've now got a simple test where just changing `get_columns` in the trino spec and making a small amendment to the base spec to respect the new optional field in `ResultSetColumnType` will make both schema previews and `SELECT *` table previews correctly query the additional nested fields.

I'll just combine my type parser work with this adjustment and the parts of this feature we most value are all complete. I'll take a look at how to make the actual data expansion work on a user query in SQL Lab as a second part to this, but imo the onyl sensible way to do that is to amend or wrap the user's query, which I worry might be misleading. I'm thinking it would actually be more sensible to provide a macro to hit this type expansion logic and give users a way of exhaustively expanding out a `ROW` field by themselves, if needed. More control to the user that way.

**Comment by giftig** [comment]
PR to implement feature: https://github.com/apache/superset/pull/25809

My type parser effort was largely wasted in the end, as it turns out someone's already contributed some very similar logic to the `trino-python-client` and it's already available in our version, but that's good news for the complexity of this feature.

I left out trying to expand the data from an arbitrary `SELECT *` or `SELECT <some_row_field>` query as there wasn't a sensible way of doing so without modifying or wrapping the query, as I described above.

So I think the above PR satisfies everything we want to see from this feature for trino.

**Reply by rusackas** [reply]
@giftig should I close out this discussion then? (Sorry to follow up so late). Thanks for the PR!

**Reply by giftig** [reply]
Ah yes this was merged a while ago, I'll close the discussion.
